axum/src/extract/ws.rs

```diff
@@ -265,9 +265,117 @@ impl<F> WebSocketUpgrade<F> {
+    /// `Sec-WebSocket-Protocol` header, no subprotocol is selected.
+    ///
+    /// This API is more flexible than [`protocols`](Self::protocols) and allows
+    /// custom selection logic, such as Token-encoded subprotocols .
+    ///
+    /// # Examples
+    ///
+    /// ## Accept a fixed protocol with no allocation
+    ///
+    /// ```rust
+    /// use http::HeaderValue;
+    /// use axum::{
+    ///     extract::ws::{WebSocketUpgrade, WebSocket},
+    ///     routing::any,
+    ///     response::{IntoResponse, Response},
+    ///     Router,
+    /// };
+    ///
+    /// let app = Router::new().route("/ws", any(handler));
+    ///
+    /// async fn handler(ws: WebSocketUpgrade) -> Response {
+    ///     ws.select_protocol(|proto| {
+    ///             (proto == "my-protocol")
+    ///                 .then(|| HeaderValue::from_static("my-protocol"))
+    ///         }).on_upgrade(|socket| async {
+    ///             // ...
+    ///         })
+    /// }
+    /// # let _: Router = app;
+    /// ```
+    ///
+    /// ## Accept token based auth
+    /// Suppose that the HTTP header is set as:
+    /// ```txt
+    /// Sec-WebSocket-Protocol: token.subprotocol.bearer.bXl0b2tlbgac7G, token.subprotocol.auth
+    /// ```
+    ///
+    /// where the `bXl0b2tlbgac7G` is the credential token
+    ///
+    /// ```rust
+    /// use http::HeaderValue;
+    /// use axum::{
+    ///     extract::ws::{WebSocketUpgrade, WebSocket},
+    ///     routing::any,
+    ///     response::{IntoResponse, Response},
+    ///     Router,
+    /// };
+    ///
+    /// let app = Router::new().route("/ws", any(handler));
+    /// async fn handler(ws: WebSocketUpgrade) -> impl IntoResponse {
+    ///     let ws = ws.select_protocol(|proto| {
+    ///         proto
+    ///             .strip_prefix("token.subprotocol.bearer.")
+    ///             .map(|token| {
+    ///                 (token == "bXl0b2tlbgac7G") // replace with real authenticate logic
+    ///                     .then_some(HeaderValue::from_static("token.subprotocol.auth"))
+    ///             })
+    ///             .flatten()
+    ///     });
+    ///     if ws.selected_protocol().is_some() {
+    ///         ws.on_upgrade(|socket| async {
+    ///             // ...
+    ///         })
+    ///     } else {
+    ///         // Reject the connection
+    ///         (axum::http::StatusCode::UNAUTHORIZED, "Unauthorized").into_response()
+    ///     }
+    /// }
+    /// # let _: Router = app;
+    /// ```
+    ///
+    /// # Notes
+    ///
+    /// - The predicate is evaluated lazily and stops at the first accepted protocol.
+    /// - The selected protocol is echoed back in the WebSocket upgrade
+    ///   response as required by RFC 6455. Some browsers may reject a
+    ///   value that was not present in the client's request.
+    /// - No allocation is performed unless the predicate constructs a dynamic
+    ///   `HeaderValue`.
```

Comment:
Why is this here? I think we generally shouldn't make any promises about allocations.